static_frame/core/util.py

```diff
@@ -3925,3 +4006,38 @@ def iloc_to_insertion_iloc(
     if key < -size or key >= size:
         raise IndexError(f'index {key} out of range for length {size} container.')
     return key % size
+
+
+def order_is_sorted_asc(arr: TNDArrayIntDefault) -> bool:
+    positions = PositionsAllocator.get(len(arr))
+
+    # If we scan the whole array without finding a mismatch, it means it is ordered
+    return first_true_1d(arr != positions, forward=True) == -1
+
+
+def _slices_from_transitions(
+    transitions: tp.Iterator[int], size: int
+) -> tp.Iterator[slice]:
+    start = 0
+    for t in transitions:
+        yield slice(start, t)
+        start = t
+
+    if start < size:
+        yield slice(start, None)
+
+
+def transition_slices_from_group(group: np.ndarray) -> tuple[tp.Iterator[slice], bool]:
+    if group.ndim == 2:
+        group_to_tuple = True
+        if group.dtype == DTYPE_OBJECT:
+            # NOTE: cannot get view of object; use string
+            consolidated = view_2d_as_1d(group.astype(str))
+        else:
+            consolidated = view_2d_as_1d(group)
+        transitions = nonzero_1d(consolidated != roll_1d(consolidated, 1))[1:]
+    else:
+        group_to_tuple = False
+        transitions = nonzero_1d(group != roll_1d(group, 1))[1:]
+
+    return _slices_from_transitions(transitions, len(group)), group_to_tuple  # type: ignore
```

Comment:
This could be implemented in ArrayKit at maybe substantial savings: integers from transitions could be read directly in C and used to create and yield slices, with all integer management in C. maybe create an issue in ArrayKit for this.